### PR TITLE
Add lang query param in findbyid endpoint

### PR DESCRIPTION
**Pull request by Joxit**
This will reduce the transfer time on the network.
We should add this in pelias API too

Closes: #42

### CODE REVIEW

**Comment by Joxit** [comment]
When the translation is not found, it will send all names [L28](https://github.com/pelias/placeholder/pull/128/files#diff-6cb3db89528bb1245eba3ab3719200ffR28).
I don't know if we should send an empty object or not.

**Comment by missinglink** [comment]
related: https://github.com/pelias/placeholder/issues/102

**Comment by missinglink** [comment]
hey @Joxit, in the PR linked above (102) @tadjik1 reports that the `JSON.parse` function is what's causing the responses to be slow.

trimming down the response body will reduce the bytes transferred over the network but might not make the actual query much faster?

File: server/routes/findbyid.js
Comment (missinglink) [inline server/routes/findbyid.js]: I forget, doesn't the 'lang' parameter need to be in a specific case (upperCase/lowerCase)?

Comment (Joxit) [reply server/routes/findbyid.js]: Yes `lang` must match the iso639-2T code in lower case

**Comment by Joxit** [comment]
Hi @missinglink, in fact this will reduce the bytes transferred and also make the query faster.

I tried the same stress test (with more requests) and I had these results :
My laptop specs 8 cores, 16Go RAM with SSD. The database contains only the objects for these id.

Without lang=fra
```
ab -n 10000 -c 100  "http://192.168.10.82:3000/parser/findbyid?ids=102063941,102191581,85682555,85633111,102064231,85682523,102063845,102063889,85687045,101807185,404473841,85633337,85682447,85633105,102064007,101875803,102064083,85688637,102081727,85927929,102191575,85633793"

Percentage of the requests served within a certain time (ms)
  50%    143
  66%    148
  75%    152
  80%    153
  90%    156
  95%    159
  98%    163
  99%    171
 100%   1169 (longest request)
```

With lang=fra
```
ab -n 10000 -c 100  "http://localhost:3000/parser/findbyid?ids=102063941,102191581,85682555,85633111,102064231,85682523,102063845,102063889,85687045,101807185,404473841,85633337,85682447,85633105,102064007,101875803,102064083,85688637,102081727,85927929,102191575,85633793&lang=fra"

Percentage of the requests served within a certain time (ms)
  50%     65
  66%     67
  75%     69
  80%     70
  90%     75
  95%     80
  98%     86
  99%     91
 100%    123 (longest request)
```

Here we can see the improvement (10x faster for the 100th percentile).

But you're right, we could also improve things by working on the database.

Something we should not forget is that on the other side there is the API that need to parse the JSON, so reduce the size here will be great anyway.

Furthermore, when placeholder send the response the JSON is again encoded to string (which also takes time if the object is large)

**Comment by missinglink** [comment]
Very interesting, it looks like this fairly simple change will have a significant improvement for performance and the HTTP API also doesn't prevent us from doing further database optimisations in the future.

File: server/routes/findbyid.js
Comment (missinglink) [reply server/routes/findbyid.js]: Can we please transform the `req.query` to lowercase, so that the interface works for both uppercase and lowercase inputs?

Comment (missinglink) [reply server/routes/findbyid.js]: Oh, oops, looks like you just coped by shoddy code :)

I think we should `toLowerCase` all the places we deal with this?

**Review by missinglink — APPROVED** [review]
I just pulled the code and tested it, I can confirm it's twice as fast on my machine using the `?lang=` param :boom:

I like that if the request `lang` is not available, then it falls back to returning all, rather than none.

Couple minor nit picks, which are probably copied from my shoddy code, could we please make some minor changes?

- lowercase inputs for `?lang=` so that users can enter the `lang` in whatever case they like, could you also please fix the `search` endpoint with the same code?
- I think `Array.isArray()` is sufficient as a null check, so `if ( lang && Array.isArray(translation) )` could probably just be `if ( Array.isArray(translation) )`, although either is fine :)

**Comment by missinglink** [comment]
oh also can we please add any relevant info to the `README` so people know how to use it? :)

**Comment by missinglink** [comment]
cc/ @orangejulius @tadjik1 any other comments before we merge this?

**Comment by Joxit** [comment]
Thank you for your review :)
The code has been updated. I have already started the changes in the API to use this.

**Comment by orangejulius** [comment]
This PR is great. It will definitely make a huge impact on performance. I'm doing some work to allow us to log/analyze our geocode.earth placeholder timing, as I think that will be the best way to see the effect in the real world.

In the meantime, how hard would it be to add unit tests for this functionality? If it would require significant work, then maybe it's not worth it right now as I don't want to hold up this significant change. But if it's relatively easy, let's definitely do it :)

**Comment by Joxit** [comment]
@orangejulius  Ok, I'm working on unit tests then :)

File: test/server/routes/findbyid.js
Comment (Joxit) [inline test/server/routes/findbyid.js:38]: Is this really expected?

Comment (missinglink) [reply test/server/routes/findbyid.js:38]: hmm.. it seems that `parseInt('1d02064231',10) === 1`

it's not very elegant, I guess we can either stick with this or just ignore any ids which contain non-numeric characters.

Comment (Joxit) [reply test/server/routes/findbyid.js:38]: Yes parseInt stops when there is a non numeric character.
Ok, so I think we should leave this like that, this will avoid useless checks.

Comment (missinglink) [reply test/server/routes/findbyid.js:38]: :+1:

**Pull request merged**
